rfcs/rfc-0000-interactions.md

```diff
@@ -0,0 +1,599 @@
+        max: Option<u32>,
+        select_type: SelectType,
+    },
+    Switch {
+        default: Option<bool>,
+    },
+    RadioButtons {
+        options: Vec<RadioButton>,
+        default: String,
+    },
+    ColourPicker {
+        transparency: bool,
+        default: u32,
+    },
+    Overflow {
+        buttons: Vec<Block>,
+    },
+}
+
+enum SelectType {
+    Member {
+        permissions: Option<PermissionsOverwrites>,
+        choices: Option<Vec<String>>,
+    },
+    Channel {
+        channel_types: Option<Vec<ChannelType>>,
+        choices: Option<Vec<String>>,
+    },
+    Role {
+        permissions: Option<PermissionsOverwrites>,
+        choices: Option<Vec<String>>,
+    },
+    Text {
+        choices: Vec<String>,
+    },
+}
+
+struct RadioButton {
+    action_id: String,
+    name: String,
+    description: Option<String>,
+}
+
+enum SelectValue {
+    String(String),
+    Number(String),
+    Attachment(String),
+    Role(String),
+    Channel(String),
+    Member(String),
+}
+
+```
+
+## Interaction Payload
+
+```rust
+struct Interaction {
+    id: String,
+
+    #[serde(flatten)]
+    inner: InteractionData,
+}
+
+
+enum InteractionData {
+    SlashCommand(InteractionDataSlashCommand),
+    Block(InteractionDataBlock),
+}
+
+struct InteractionDataBlock {
+    action_id: String,
+
+    #[serde(flatten)]
+    inner: InteractionDataBlockInner,
+}
+
+enum InteractionDataBlock {
+    SubBlock {
+        blocks: Vec<InteractionDataBlock>,
+    },
+    Button,
+    CheckBox {
+        value: bool,
+    },
+    TextInput {
+        value: String,
+    },
+    Attachment {
+        value: String,
+    },
+    Number {
+        value: String,
+    },
+    Select {
+        value: SelectValue,
+    },
+    RadioButtons {
+        value: String
+    },
+    ColourPicker {
+        value: u64
+    },
+}
+
+struct InteractionDataSlashCommand {
+    name: String,
+    id: String,
+    options: Vec<InteractionDataSlashCommandOption>,
+}
+
+struct InteractionDataSlashCommandOption {
+    name: String,
+
+    #[serde(flatten)]
+    inner: InteractionDataSlashCommandOptionInner,
+}
+
+enum InteractionDataSlashCommandOptionInner {
+    Text {
+        value: String,
+    },
+    Number {
+        value: String,
+    },
+    Boolean {
+        value: bool,
+    },
+    Member {
+        member: Member,
+        user: User,
+    },
+    Channel {
+        channel: Channel,
+    },
+    Role {
+        role: Role,
+    },
+    Date {
+        value: String,
+    },
+    Time {
+        value: u32,
+    },
+    Datetime {
+        date: String,
+        time: u32,
+    },
+    Attachment {
+        value: String,
+    },
+    VarArgs {
+        values: Vec<InteractionDataSlashCommandOptionInner>,
+    },
+    Group {
+        group: InteractionDataSlashCommandOption,
+    },
+}
+```
+
+## Responding To Interactions
+
+After receiving an interaction the bot has 3 seconds to respond to the interaction with a reponse, this
```

Comment:
could this be bumped up to at least 5 seconds? i'd be worried that response times could occasionally exceed this limit.

additionally, shouldn't there be a way to defer responses for operations that can possibly take a full minute or longer to complete?

Reply:
Bumping it a couple seconds should be fine, i don't like deferring as it doesn't give much feedback to the user, i could possible add it but it needs to be clear that there is still another response coming, this could be done via just editing the response but i cant think of a good way to handle this at the moment.

Reply:
defer would be useful

Reply:
We dont need defers since we can post a pending message like "A response is incoming" by the bot itself and edit the message once the command is done, but if the defers are to implemented, could we possibly get a custom message for it?

Reply:
I think having a message or an option to mark as "not finished" or "loading" would be the best solution as it makes it clear to the user that something is coming rather.

I will update the RFC soon to reflect my idea.

Reply:
> Bumping it a couple seconds should be fine, i don't like deferring as it doesn't give much feedback to the user, i could possible add it but it needs to be clear that there is still another response coming, this could be done via just editing the response but i cant think of a good way to handle this at the moment.

Why not have a similar aproach to discord where a placeholder text a la "<bot> is thinking..." is given?
A defer has its benefits here imo